euphonic_wrapper.py

```diff
@@ -0,0 +1,194 @@
+import numpy as np
+import os
+from euphonic import Crystal, ForceConstants, QpointPhononModes, DebyeWaller, ureg
+from euphonic.util import mp_grid
+
+class EuphonicWrapper(object):
+    # This a wrapper around the Euphonic ForceConstants and QpointPhononModes classes to make it easier to access from Matlab
+    # It is meant to be used with a Matlab python_wrapper class and implements a horace_sqw function for use with Horace
+
+    defaults = {'force_constants': None, 'phonon_modes': None, 'debye_waller': None, 'debye_waller_grid': None, 
+                'temperature': 0.0 * ureg('K'), 'bose': True, 'negative_e': False, 'conversion_mat': None, 'chunk': 5000, 
+                'lim': np.inf, 'scattering_lengths': 'Sears1992', 'weights': None, 'asr': None, 'dipole': True, 
+                'eta_scale': 1.0, 'splitting': True, 'insert_gamma': False, 'reduce_qpts': True, 'use_c': False, 
+                'n_threads': 1, 'fall_back_on_python': True, 'convolution_function': 'gauss'}
```

Comment:
Yeah, not yet - I was doing too many things at once. So at the moment to use this with `multifit` / `tobyfit` you have to use `disp2sqw` to wrap it, because both of those send the 4-vector (h,k,l,en) to the model function and expect a single vector S(q,w) back - so for Euphonic we need to do the energy convolution which `disp2sqw` does... An example syntax is:

```matlab
mf = multifit_sqw(wsc);
mf = mf.set_fun(@disp2sqw, {@euobj.horace_sqw, {scalefac}, [fwhm]});
wsim_mf = kk.simulate('fore');
```

I was thinking of implementing this directly in Python, so you can use:

```matlab
mf = multifit_sqw(wsc);
mf = mf.set_fun(@euobj.horace_sqw, {[scalefac fwhm]});
```

But this is probably not necessary (and is throwing up an issue with passing an energy dependent FWHM). 

I'll just remove the `convolution_function` argument.

Reply:
Thanks, this is useful info.

> at the moment to use this with `multifit` / `tobyfit`

Are there plans to change this so multifit works with (h,k,l) functions? It seems like it would be useful for the multifit/tobyfit to be able to parameterise and apply its own broadening... the broadening in Euphonic isn't that sophisticated.

If we do need a function that returns the binned/broadened sqw, it should probably be a separate method. Does the name `horace_sqw` imply that binning/broadening has already been applied? Maybe we should rename things and have `coherent_model.horace_disp`  that returns `w, sf` and `coherent_model.horace_sqw` that returns binned/broadened `sqw`? Although the `coherent_model.horace_sqw` doesn't have to be implemented in this PR, it helps if we make sure we have our naming sorted out for future changes. Please let me know if you disagree with any of these things or the naming ideas I've just thrown out there...

Reply:
Yeah, it's probably best to rename it to `horace_disp` rather than `horace_sqw`. I'll do that, but leave `horace_sqw` unimplemented 

(for now... I have an implementation which should be faster than doing it in Matlab. This is because `disp2sqw` expects a cell array of vectors for `w` and `sf` whereas we just compute a 2D numpy array. To get this into Matlab I have to make a transposed copy in Python, map this to a `tuple` and then in the interface code make another copy to a Matlab `CellArray` because there's an issue with using memory across library boundaries (it's a limitation of the Matlab-Python implementation I think). So that's two copies of an `n_qpts x n_modes` array rather than a single copy of a `1 x n_qpts` vector if we did the energy convolution in Python...

The issue is that I wanted to replicate the functionality of `disp2sqw` which allows you to get a function handle for the `fwhm` parameter which allows you to input an energy-dependent FWHM which is what David wants. Unfortunately if you're using Matlab you cannot pass a Matlab function handle to Python. (for PyHorace you'd pass a Python function reference so it's fine). There's a way around this with a compiled C++ `mex` / CPython module but might be messy... for this specific case I think I can refactor the Matlab wrapper to compute the energy dependent fwhm for the given input energy transfer vector and just pass that to Python as a vector... Anyhow a bit more exploration needed...